control_plane/attachment_service/src/service.rs

```diff
@@ -143,31 +144,71 @@ impl Service {
+        {
+            let nodes = {
+                let locked = self.inner.read().unwrap();
+                locked.nodes.clone()
+            };
+            for node in nodes.values() {
+                let http_client = reqwest::ClientBuilder::new()
+                    .timeout(Duration::from_secs(5))
+                    .build()
+                    .expect("Failed to construct HTTP client");
+                let client = mgmt_api::Client::from_client(
+                    http_client,
+                    node.base_url(),
+                    self.config.jwt_token.as_deref(),
+                );
 
-            tracing::info!("Scanning shards on node {}...", node.id);
-            match client.list_location_config().await {
-                Err(e) => {
-                    tracing::warn!("Could not contact pageserver {} ({e})", node.id);
-                    // TODO: be more tolerant, apply a generous 5-10 second timeout with retries, in case
-                    // pageserver is being restarted at the same time as we are
+                fn is_retryable(e: &mgmt_api::Error) -> bool {
+                    use mgmt_api::Error::*;
+                    match e {
+                        ReceiveBody(_) | ReceiveErrorBody(_) => true,
+                        ApiError(StatusCode::SERVICE_UNAVAILABLE, _)
+                        | ApiError(StatusCode::GATEWAY_TIMEOUT, _)
+                        | ApiError(StatusCode::REQUEST_TIMEOUT, _) => true,
+                        ApiError(_, _) => false,
+                    }
                 }
-                Ok(listing) => {
-                    tracing::info!(
-                        "Received {} shard statuses from pageserver {}, setting it to Active",
-                        listing.tenant_shards.len(),
-                        node.id
-                    );
 
-                    for (tenant_shard_id, conf_opt) in listing.tenant_shards {
-                        observed.insert(tenant_shard_id, (node.id, conf_opt));
+                let list_response = backoff::retry(
+                    || client.list_location_config(),
+                    |e| !is_retryable(e),
+                    1,
+                    5,
+                    "Location config listing",
+                    &cancel,
+                )
+                .await;
+                let Some(list_response) = list_response else {
+                    tracing::info!("Shutdown during startup_reconcile");
+                    return;
+                };
+
+                tracing::info!("Scanning shards on node {}...", node.id);
+                match list_response {
+                    Err(e) => {
+                        tracing::warn!("Could not contact pageserver {} ({e})", node.id);
+                        // TODO: be more tolerant, do some retries, in case
+                        // pageserver is being restarted at the same time as we are
```

Comment:
Does this result in loss of availability or should computes already be configured to point to the pageserver we couldn't contact?

Reply:
If we can't contact a pageserver on startup, we will omit any content for that pageserver in the intent+observed states, causing schedule() and maybe_reconcile() respectively to do some work (e.g. find a new home for the tenant).  This is potentially expensive, so this code path would benefit from a bit more nuance on the handling, e.g. if we see _all_ pageservers offline, then we should wait rather than proceeding in a state where we wouldn't be able to schedule anyone.

From the compute's point of view, if they're currently talking to a pageserver that hits this "Could not contact" path, if it's responding to them (but not to the storage controller) it'll keep working for page requests, and they'll just get a config update to talk to a different pageserver once we've been through the schedule->reconcile process to attach the shard somewhere different.  

The missing piece currently is to clean up a pageserver when it eventually comes back online: that will be added later to the path were we set a node to Active state: upon that change, we should reconcile with the node.